packages/remix-dev/cli/commands.ts

```diff
@@ -234,3 +237,103 @@ export async function codemod(
     throw error;
   }
 }
+
+export async function generateEntry(remixRoot: string, entry: string) {
+  // 1. validate requested entry file
+  let clientEntries = new Set([
+    "entry.client.tsx",
+    "entry.client.js",
+    "entry.client.jsx",
+  ]);
+  let serverEntries = new Set([
+    "entry.server.tsx",
+    "entry.server.js",
+    "entry.server.jsx",
+  ]);
+
+  let entries = new Set([...clientEntries, ...serverEntries]);
+
+  if (!entries.has(entry)) {
+    // @ts-expect-error available in node 12+
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat#browser_compatibility
+    let listFormat = new Intl.ListFormat("en", {
+      style: "long",
+      type: "conjunction",
+    });
+
+    let entriesArray = Array.from(entries);
+    let list = listFormat.format(entriesArray);
+
+    console.log(
+      colors.error(`Invalid entry file. Valid entry files are ${list}`)
+    );
+    return process.exit(1);
+  }
+
+  let defaultsDirectory = path.resolve(__dirname, "..", "config", "defaults");
+  let defaultEntryClient = path.resolve(defaultsDirectory, "entry.client.tsx");
+  let defaultEntryServer = path.resolve(defaultsDirectory, "entry.server.tsx");
+
+  // 2. check if any of the requested entry file exists
+  let entryType = entry.startsWith("entry.client.") ? "client" : "server";
+  let inputFile =
+    entryType === "client" ? defaultEntryClient : defaultEntryServer;
+  let outputFile = path.resolve(remixRoot, "app", entry);
+
+  let entriesToCheck = entryType === "client" ? clientEntries : serverEntries;
+  for (let entryToCheck of entriesToCheck) {
+    let entryExists = await fse.pathExists(
+      path.resolve(remixRoot, "app", entryToCheck)
+    );
+    if (entryExists) {
+      console.log(
+        colors.gray(
+          `Entry file ${path.relative(remixRoot, entryToCheck)} already exists.`
+        )
+      );
+      return process.exit(1);
+    }
+  }
+
+  let contents: string | undefined;
+
+  // 3. if server entry, update runtime import
+  if (entryType === "server") {
+    contents = await fse.readFile(inputFile, "utf-8");
+
+    let pkgJson = await NPMCliPackageJson.load(remixRoot);
+    let deps = pkgJson.content.dependencies ?? {};
+
+    if (deps["@remix-run/node"]) {
+      // we good
+    } else if (deps["@remix-run/cloudflare"]) {
+      contents = contents.replace(/@remix-run\/node/g, "@remix-run/cloudflare");
+    } else if (deps["@remix-run/deno"]) {
+      contents = contents.replace(/@remix-run\/node/g, "@remix-run/deno");
+    }
+  }
+
+  // 3. if entry is js/jsx, convert to js
+  // otherwise, copy the entry file from the defaults
+  if (/\.jsx?$/.test(entry)) {
+    contents = contents ||= await fse.readFile(inputFile, "utf-8");
```

Comment:
I think you meant

```suggestion
    contents ||= await fse.readFile(inputFile, "utf-8");
```

Reply:
yes, yes i did 🤪 - gonna need to also update PassThrough in deno/cf too, so may go with something different anyways